modules/katello_repository.py

```diff
@@ -175,42 +189,36 @@ def main():
             content_type=dict(required=True),
             url=dict(),
             download_policy=dict(choices=['background', 'immediate', 'on_demand']),
+            state=dict(required=True, choices=['present_with_defaults', 'present', 'absent']),
         ),
         supports_check_mode=True,
     )
 
     handle_no_nailgun(module, HAS_NAILGUN_PACKAGE)
 
-    server_url = module.params['server_url']
-    verify_ssl = module.params['verify_ssl']
-    username = module.params['username']
-    password = module.params['password']
-    product = module.params['product']
-    organization = module.params['organization']
-    name = module.params['name']
-    content_type = module.params['content_type']
-    url = module.params['url']
-    download_policy = module.params['download_policy']
-
-    server = ServerConfig(
-        url=server_url,
-        auth=(username, password),
-        verify=verify_ssl
-    )
-    ng = NailGun(server, entities, module)
+    entity_dict = dict(
+        [(k, v) for (k, v) in module.params.items() if v is not None])
+
+    server_url = entity_dict.pop('server_url')
+    verify_ssl = entity_dict.pop('verify_ssl')
+    username = entity_dict.pop('username')
+    password = entity_dict.pop('password')
+    state = entity_dict.pop('state')
 
-    # Lets make an connection to the server with username and password
     try:
-        org = entities.Organization(server)
-        org.search()
+        create_server(server_url, (username, password), verify_ssl)
     except Exception as e:
         module.fail_json(msg="Failed to connect to Foreman server: %s " % e)
 
-    try:
-        changed = ng.repository(name, content_type, product, organization, url=url, download_policy=download_policy)
-        module.exit_json(changed=changed)
-    except Exception as e:
-        module.fail_json(msg=e)
+    ping_server(module)
+
+    entity = find_repository(module, name=entity_dict['name'], product=entity_dict['product'])
```

Comment:
I think, your runtime error is based here. `entity_dict['product']` is the name of the product at this point while it should be a nailgun entity object. I think, you need a `find_product` first.

Reply:
Any reason why we dont let the find objects do these lookups for us to minimize effort in the modules themselves?

Reply:
we need to have them resolved to nailgun objects for the naildown_state.